bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java

```diff
@@ -449,20 +455,61 @@ private static boolean isConvertibleToArrayType(Object sourceValue, BArrayType t
                 return true;
             }
         }
-        Set<Type> convertibleTypes;
+
         for (int i = 0; i < source.size(); i++) {
-            convertibleTypes = getConvertibleTypes(source.get(i), targetTypeElementType, unresolvedValues);
-            if (convertibleTypes.isEmpty()) {
+            if (!isConvertibleToArrayInstance(source.get(i), targetTypeElementType, unresolvedValues)) {
                 return false;
             }
-            if (convertibleTypes.size() != 1 && !convertibleTypes.contains(TypeChecker.getType(source.get(i)))
-                    && !hasIntegerSubTypes(convertibleTypes)) {
+        }
+        return true;
+    }
+
+    private static boolean isConvertibleToTupleType(Object sourceValue, BTupleType targetType,
+                                                    List<TypeValuePair> unresolvedValues) {
+        if (!(sourceValue instanceof ArrayValue)) {
+            return false;
+        }
+
+        ArrayValue source = (ArrayValue) sourceValue;
+        List<Type> targetTypes = targetType.getTupleTypes();
+        int sourceTypeSize = source.size();
+        int targetTypeSize = targetTypes.size();
+        Type targetRestType = targetType.getRestType();
+
+        if (sourceTypeSize < targetTypeSize) {
+            return false;
+        }
+        if (targetRestType == null && sourceTypeSize > targetTypeSize) {
+            return false;
+        }
+
+        for (int i = 0; i < targetTypeSize; i++) {
+            if (!isConvertibleToArrayInstance(source.getRefValue(i), targetTypes.get(i), unresolvedValues)) {
+                return false;
+            }
+        }
+
+        for (int i = targetTypeSize; i < sourceTypeSize; i++) {
+            if (!isConvertibleToArrayInstance(source.getRefValue(i), targetRestType, unresolvedValues)) {
                 return false;
             }
         }
         return true;
     }
 
+    private static boolean isConvertibleToArrayInstance(Object sourceElement, Type targetType,
+                                                        List<TypeValuePair> unresolvedValues) {
+        Set<Type> convertibleTypes = getConvertibleTypes(sourceElement, targetType, unresolvedValues);
+        if (convertibleTypes.isEmpty()) {
+            return false;
+        }
+        if (convertibleTypes.size() > 1 && !convertibleTypes.contains(TypeChecker.getType(sourceElement))
+                && !hasIntegerSubTypes(convertibleTypes)) {
+            return false;
+        }
```

Comment:
I think the following is possible
```suggestion
        if (convertibleTypes.isEmpty() || (convertibleTypes.size() > 1 && !convertibleTypes.contains(TypeChecker.getType(sourceElement))
                && !hasIntegerSubTypes(convertibleTypes))) {
            return false;
        }
```
We can move the boolean checks to a separate function. I mean what is inside the brackets of the `if` statement

Reply:
Yeah, it is possible, but we are going to create too many instructions for a failing case right? I was just thinking about for an early return when the convertible types set is empty. WDYT?